arcaflow_plugin_fio/fio_plugin.py

```
@@ -52,9 +56,10 @@ def run(
             error_output: FioErrorOutput = FioErrorOutput(format_exc())
         return "error", error_output
 
-    except Exception:
```

Do you think it would be a good idea to keep the `except Exception` in case a specific exception is left unhanded, or do you think every possible exception possibility is accounted for?

---

> do you think every possible exception possibility is accounted for?

I don't see how that could possibly be true.  🙃

The question is, if an unexpected exception is encountered here, what if any response should this code have.  What the code was doing previously looks good, so perhaps it should be retained.  However, if it's redundant with what will happen with an unhandled exception after the `finally` clause completes, then it would probably be best to omit it.  I don't know whether it is, but it would be annoying to have the failure mode "eaten" because we didn't send the traceback to the right place.

---

Let me ask somewhat naively, is it best practice to have a catch-all `Exception` like this? I don't think we've done this consistently across the project, and Pylint has a habit of complaining about this: https://pylint.readthedocs.io/en/latest/user_guide/messages/warning/broad-exception-caught.html

---

Generally "unexpected" exceptions should be unhandled; but that depends on having some outer scope that'll report them and abort in some defined way. (E.g., Flask and FastAPI will generate a server stack trace and return an HTTP 500 error with the exception text.)

So the question(s) *here* might be:

1. What does the ARCA SDK wrapper do if this raises some other exception? Does the plugin fail gracefully with some meaningful indication of failure? Is there a log to examine?
2. Is there any *specific* context-specific action we might want to take *here* when something else happens? There are places where it makes sense to issue a context-dependent log message and then re-raise, for example; and others where we might want to log and then ignore the exception.

---

> is it best practice to have a catch-all `Exception` like this?

As Dave intimated, the answer to that is context-sensitive:
- As a general rule, code should only _catch_ (i.e., attempt to recover from) exceptions that it recognizes.  It is unreasonable to expect code to be able to correctly recover from an error which it does not properly anticipate -- such errors should either be left unhandled or properly re-raised.
- That said, code should probably attempt to restore any local invariant which it has affected before letting the exception propagate, but this would typically be done with a `finally` clause, not an `except` clause, so it's not really "catching" the exception.  (And, note that this may be dangerous to attempt, given that the code doesn't understand the nature of the failure.)
- However, depending on the SLA/requirements of the code as a whole, it may not be appropriate for it to crash in the event of an unhandled exception.  Likewise, depending on the environment in which the tool is running, it would be counterproductive to allow details of the crash to be lost.  So, at or near the "external interface" of the code (whether that's a UI or a library API), it might be appropriate to have a "catch all" handler which appropriately logs/reports the details of the error and then attempts to recover or restore operation.  But, this is really a unique exception to the rule.

So, I second Dave's questions:
- is there some handler up the stack from this one which will take care of reporting this error (so that it is redundant to do so here)?
- will the plugin still react appropriately (e.g., fail gracefully) if we don't handle the error here?
- is there any specific action which would be inconvenient to implement elsewhere which we should take _here_ in order to be helpful (such as logging local context or information)?

---

The problem with the broad catch-all is if it's used to hide problems. In this case it's reporting an error. So as long as it's done properly, I'd leave it in.
However, in a more recent version of the SDK I included a catch-all in the handler function that calls the function in the plugin. Its handling of it won't be the most graceful, but it will at least report failure of the step with a copy of the error. So it's not too bad if it's left out. Before this change, leaving that out would mean that there would be no insights into what caused the failure.

---

> there would be no insights into what caused the failure.

If the up-stack handler logs the traceback, then it should include the location of the failure.  Unless something logs more than that, it's probably best to have it logged as far up the stack as feasible.

On the flip side, we don't need to have the traceback logged more than once....

---

So I can tell you in the particular case that led to this change, the plugin was being run via the Arcaflow engine and the exception was raised to the engine's debug log output, but a formatting problem stopped it from showing the content of the error from the underlying `subprocess`. Now I suppose that has nothing to do with whether that exception was caught as a generic `Exception` or as a `subprocess.CalledProcessError` and really the underlying issue was in not capturing the text output of the error correctly. It made sense to me while fixing this to also make the exception more explicit.

But if I'm not mistaken (based only on fuzzy memory of similar situations), an unhandled exception from a plugin is raised in a similar way to the engine's debug logs. I'm not sure there is any real difference in that case -- The plugin crashes (we probably want it to), and a traceback is reported. Does adding the generic Exception change that?

---

The exception handling that I added should be good enough if we want to leave it out.